### ISSUE TITLE
Reproducing BioASQ Results

### ISSUE THREAD

**Issue by suamin** [dmis-lab/biobert#34]
Hi,
First of all, thanks for the BioBERT models.
I was trying to reproduce the results of BioASQ 6b with BioBERT (+PubMed + PMC) as mentioned in Table 6. which states `S=35.58 / L=51.39 / M=42.51`. I'm using hugging face's [run_squad.py](https://github.com/huggingface/pytorch-pretrained-BERT/blob/master/examples/run_squad.py) script. I tried two settings, one without fine-tuning pre-trained model on SQuAD and other with fine-tuning. 

For SQuAD training I used this command:
```
export BERT_MODEL=/path/to/pubmed_pmc_470k
export SQUAD_DIR=/path/to/squad

python run_squad.py \
  --bert_model $BERT_MODEL \
  --do_train \
  --do_predict \
  --train_file $SQUAD_DIR/train-v1.1.json \
  --predict_file $SQUAD_DIR/dev-v1.1.json \
  --train_batch_size 6 \
  --learning_rate 3e-5 \
  --num_train_epochs 2.0 \
  --max_seq_length 384 \
  --doc_stride 128 \
  --output_dir /path/to/squad_tuned_output
```
Then, I used fine-tuned model to train on 6b-train.
```
export BERT_MODEL=/path/to/squad_tuned_output
export BIOBERT_BIOASQ_DIR=/path/to/biobert-bioasq-datafiles

python run_squad.py \
  --bert_model $BERT_MODEL \
  --do_train \
  --train_file $BIOBERT_BIOASQ_DIR/BioASQ-train-factoid-6b.json \
  --train_batch_size 6 \
  --learning_rate 5e-6 \
  --num_train_epochs 5.0 \
  --max_seq_length 384 \
  --doc_stride 128 \
  --output_dir /path/to/output
```
Below I summarize results for each test batch of 6b (1-5) for both cases:

Without SQuAD fine-tuning
```
  S     L     M 
29.03 54.83 39.08 | 6b1
28.57 66.66 45.63 | 6b2
37.50 56.25 43.54 | 6b3
27.27 48.48 33.58 | 6b4
36.36 50.00 40.90 | 6b5
------------------------
31.74 55.24 40.54 | Avg
```
With SQuAD fine-tuning
```
  S     L     M 
29.03 64.51 43.38 | 6b1
38.09 76.19 52.77 | 6b2
46.87 62.50 53.12 | 6b3
39.39 48.48 42.17 | 6b4
38.63 52.27 43.71 | 6b5
-----------------------
38.40 60.79 47.03 | Avg
```
Assuming paper is reporting results with average on all 5 batches as well then the results I'm able to reproduce are giving a boost of **+2.82 / +9.4 / +4.52** for S, L and M respectively.

Would you comment on this, do you think its due to some implementation differences between TensorFlow and PyTorch implementations? or if you see an issue with the way I reproduced the results. Imo, the difference between results is very high for L and M and If you're able to reproduce the results, that I have reported, then maybe it'll be helpful to update arXiv version of paper for other researchers working on same problems to collect the right SOTA. 

Thanks!

**Comment by wonjininfo** [comment]
Hi, 
We updated code, data and released BioBERT v1.1 on 16th of May. 
I believe the difference is caused by this update of dataset. 
We will update arXiv version in the near future. 

Before we release new version on arXiv, I will share our results here. 

BioASQ 6b (With SQuAD fine-tuning )
```
BioBERT v1.0 (+PubMed + PMC) : S 40.37 L 57.77 M 47.48 (Micro average)
```
and
```
BioBERT v1.1 (+PubMed) : S 42.86 L 57.77 M 48.43 (Micro average)
```

Thanks!

**Comment by suamin** [comment]
Hi @wonjininfo 

Thanks for the prompt comment.

I missed the update and had v1.0 model, while working with newer dataset. With your, newly posted, results, the difference is relatively less. Did the update changed results for 4b and 5b as well and lastly, what has changed between v1.0 to v1.1 models?

Thanks again and great results!

**Comment by suamin** [comment]
Hi @wonjininfo,
Can you provide the script or elaborate the steps that you apply to convert original BioASQ 6b dataset (train + test) to the form you're distributing. The BioASQ 6b training data has 619 factoid questions with multiple context documents, since not all the documents contain answers therefore one can enumerate pairs of `<question, document containing answer as context>` to create the dataset as required in SQuAD format. The training data as provided by BioBERT has 4772 such pairs while when I try to reproduce, I'm getting around 13k. Also, do you create context by combining a PubMed article with its title and abstract? 
It will be very helpful if you can provide these conversion scripts or mention the steps. Thank you!

**Comment by wonjininfo** [comment]
Hi @suamin , 
Yes we created context by combining a PubMed article with its title and abstract. 
I'm afraid to tell you that we decided not to distribute our pre-processing code. 

I think the main cause of the number difference between our data and your pre-processed data lies in following settings.
> Instead of setting answer locations as every exact matching words, snippets field of original dataset are used to find proper answer location.
> We search for the offset of the first exact matching words in snippet, and add it to the offset of the snippet in the paragraph.
> In this way, we can find plausible location of answer within the paragraph. 

cf) These settings will be presented in our oncoming paper on BioASQ challenge (expected within 1~3 weeks).

**Comment by suamin** [comment]
Thanks for the insights and good luck with the challenge. Just saw the leaderboard, great results :)

**Comment by suamin** [comment]
Zero-shot predictions on test batches of 6b using BioBERT with SQuAD 1.1 fine-tuning only (**no training on 6b**) gives almost equal results as with training:

```
  S      L      M     
32.26  61.29  45.27 | 6b1
52.38  66.67  56.90 | 6b2
43.75  65.63  51.93 | 6b3
36.36  48.48  40.51 | 6b4
31.82  56.82  42.58 | 6b5
-------------------------
39.31  59.77  47.43 | Avg
```
Can you comment if you noted similar?

**Comment by wonjininfo** [comment]
Yes we are aware of it. 
As long as I remember, fine-tuning on BioASQ datasets adds a relatively small (but still meaningful) improvement on scores. Since the size of each test dataset is quite small, I suggest you use all three of datasets (4b/5b/6b). 
Thanks!

**Comment by suamin** [comment]
Thanks!

As can be seen, lenient accuracy is 2 points higher than both of v1.0/1.1 results (posted by you). For MRR, only v1.1 is slightly better. However, strict accuracy is almost +2 points for your v1.1 but if we look at results reported in paper (`S=35.58 / L=51.39 / M=42.51`) then zero-shot is much higher. Also, in my experiments, too much training actually degrades the performance than zero-shot but I will further probe the results and do full training. 

Sorry I didn't get your suggestion; 6b training includes test sets of 4b/5b/earlier. While BioASQ datasets are small, it should be recalled that SQuAD is 100k and we train it for 2 epochs which, ideally, should be good enough to answer any factoid based questions. 

On the other side, it makes for good argument that SQuAD fine-tuning is enough for understanding BioASQ QA task and further training is not meaningful. What will be more interesting to see is, how the performance varies between original BERT and BioBERT (both) with SQuAD fine-tuning in a zero-shot setting (in your paper BERT results were mentioned but I'm not sure if SQuAD fine-tuning was performed?); if former does better or equally good then, one can argue for BERT's robustness to domain-shifts (i.e. require no (zero-shot) or minimal data to adapt to new domain) which, interestingly, will contradict the recent findings of Yogatama et al. 2019 (https://arxiv.org/pdf/1901.11373.pdf) or makes point for being over-fitted to SQuAD quirks (or questions the quality of BioASQ dataset).

**Comment by abdallah197** [comment]
Hi @suamin, what did you use to run the evaluation on the test files?

**Comment by suamin** [comment]
Hi @abdallah197 , to run the official BioASQ evaluation, you will be required to download the `Evaluation-Measures` toolkit provided by them. To reproduce numbers you will need `pytorch-transformers` and the `biobert` repository. Assuming they all have been cloned into a dir `clone` then proceed as follows:

```
BIOBERT_BIOASQ_DIR=biobert_bioasq_data_dir_task6b
OUTPUT_DIR=trained_model_output_dir
TEST_BATCH=1 # change batch number here from 1 to 5

python clones/pytorch-transformers/examples/run_squad.py \
    --model_type bert \
    --model_name_or_path $OUTPUT_DIR \
    --do_eval \
    --train_file $BIOBERT_BIOASQ_DIR/BioASQ-train-factoid-6b.json \
    --predict_file $BIOBERT_BIOASQ_DIR/BioASQ-test-factoid-6b-$TEST_BATCH.json \
    --learning_rate 5e-6 \
    --num_train_epochs 5 \
    --max_seq_length 384 \
    --doc_stride 128 \
    --output_dir $OUTPUT_DIR \
    --per_gpu_eval_batch_size=3   \
    --per_gpu_train_batch_size=3 \
    --save_steps 1000000

PREDICTIONS_DIR=$OUTPUT_DIR/predictions
mkdir $PREDICTIONS_DIR
mv $OUTPUT_DIR/predictions_.json $PREDICTIONS_DIR/predictions_6b_$TEST_BATCH.json
mv $OUTPUT_DIR/nbest_predictions_.json $PREDICTIONS_DIR/nbest_predictions_6b_$TEST_BATCH.json

python clones/biobert/biocodes/transform_nbset2bioasqform.py --nbest_path=$PREDICTIONS_DIR/nbest_predictions_6b_$TEST_BATCH.json --output_path=$PREDICTIONS_DIR

mv $PREDICTIONS_DIR/BioASQform_BioASQ-answer.json $PREDICTIONS_DIR/BioASQform_BioASQ-answer-6b-$TEST_BATCH.json
mv $PREDICTIONS_DIR/WithProb_BioASQform_BioASQ-answer.json $PREDICTIONS_DIR/WithProb_BioASQform_BioASQ-answer-$TEST_BATCH.json

java -Xmx10G -cp $CLASSPATH:./clones/Evaluation-Measures/flat/BioASQEvaluation/dist/BioASQEvaluation.jar evaluation.EvaluatorTask1b -phaseB -e 5 \
    $BIOBERT_BIOASQ_DIR/6B$TEST_BATCH\_golden.json \
    $PREDICTIONS_DIR/BioASQform_BioASQ-answer-6b-$TEST_BATCH.json
```
